tests/test_app.py

```diff
@@ -1,94 +1,152 @@
 import hashlib
+from datetime import datetime, timedelta, timezone
+from urllib.parse import urljoin
 
-import pytest
-from fastapi import HTTPException
+import itsdangerous
 from fastapi.testclient import TestClient
 
-from hatch import app, config, models
+from hatch import app, config, schema, signing
 
 
 client = TestClient(app.app)
 
 
-def test_validate_raises_with_bad_secret():
-    with pytest.raises(HTTPException):
-        app.validate("bad")
+def auth_token(path, user="user", expiry=None, scope="view"):
+    if expiry is None:  # pragma: no cover
+        expiry = datetime.utcnow() + timedelta(hours=1)
 
+    return signing.AuthToken(
+        url=urljoin(client.base_url, path),
+        user=user,
+        expiry=expiry,
+        scope=scope,
+    )
+
+
+def auth_headers(workspace="workspace", user="user", expiry=None, scope="view"):
+    """Helper to create valid authentication headers for a specific workspace"""
+    token = auth_token(f"/workspace/{workspace}", user, expiry, scope)
+    return {"Authorization": token.sign()}
 
-def test_validate_succeeds():
-    assert app.validate("secret") is None
+
+def test_validate_invalid_token():
+    url = "/workspace/workspace/current"
+    token = auth_token(url)
+    # we can not easily call validate() directly, as fastapi's Request object
+    # is very much not sans-io, and thus difficult just instantiate.
+    serializer = itsdangerous.Signer("badsecret")
+    response = client.get(url, headers={"Authorization": token.sign(serializer)})
+    assert response.status_code == 403
+
+
+def test_validate_url():
+    headers = auth_headers()
+    r1 = client.get("/workspace/workspace/current", headers=headers)
+    # this means it was valid, but workspace did not exist
+    assert r1.status_code == 404
+
+    # test file url prefix
+    r2 = client.get("/workspace/workspace/current/output/file.txt", headers=headers)
+    # this means it was valid, but workspace did not exist
+    assert r2.status_code == 404
+
+    # test file url workspace url
+    r3 = client.get("/workspace/other/current/output/file.txt", headers=headers)
+    assert r3.status_code == 403
```

Comment:
On the last two asserts you explain why it returned that code, it was really useful! Could you add that here too?